tests/cpptests/iterator_util.h

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright Redis Ltd. 2016 - present
+ * Licensed under your choice of the Redis Source Available License 2.0 (RSALv2) or
+ * the Server Side Public License v1 (SSPLv1).
+ */
+
+#include "src/iterators/iterator_api.h"
+
+#include <stdint.h>
+#include <vector>
+#include <stdexcept>
+
+extern "C" {
+    IteratorStatus MockIterator_Read(QueryIterator *base);
+    IteratorStatus MockIterator_SkipTo(QueryIterator *base, t_docId docId);
+    size_t MockIterator_NumEstimated(QueryIterator *base);
+    void MockIterator_Rewind(QueryIterator *base);
+    void MockIterator_Free(QueryIterator *base);
+}
+
+class MockIterator {
+public:
+    QueryIterator base;
+    std::vector<t_docId> docIds;
+    size_t nextIndex;
+    IteratorStatus whenDone;
+    size_t readCount;
+private:
+
+    static void setBase(QueryIterator *base) {
+        base->type = READ_ITERATOR;
+        base->isValid = true;
+        base->lastDocId = 0;
+        base->current = NewVirtualResult(1, RS_FIELDMASK_ALL);
+        base->NumEstimated = MockIterator_NumEstimated;
+        base->Free = MockIterator_Free;
+        base->Read = MockIterator_Read;
+        base->SkipTo = MockIterator_SkipTo;
+        base->Rewind = MockIterator_Rewind;
+    }
+public:
+    // Public API
+    IteratorStatus Read() {
+        readCount++;
+        if (nextIndex >= docIds.size() || !base.isValid) {
+            base.isValid = false;
+            return whenDone;
+        }
+        base.lastDocId = base.current->docId = docIds[nextIndex++];
+        return ITERATOR_OK;
+    }
+    IteratorStatus SkipTo(t_docId docId) {
+        readCount++;
```

Comment:
why can't it be at this scope and remove the decrement? is this due to checking exceptions?
```suggestion
        readCount++;
```

Reply:
Yes, exceptions and EOF check